### PR TITLE
[SMTChecker] Output values for block, msg and tx variables in cex

### LINKED ISSUE #10785: [SMTChecker] CHC Show tx variables in counterexample
`msg.value` is implemented in https://github.com/ethereum/solidity/pull/10828 when > 0, but it would be nice to also implement the others, when they're used.
An idea is, for every function in the transaction trace, if a `block`, `msg` or `tx` variable is used (can be detected syntactically), show its value in the transaction function call.

### PR DESCRIPTION
**Pull request by leonardoalt**
Fix https://github.com/ethereum/solidity/issues/10785
Depends on https://github.com/ethereum/solidity/pull/12036

This PR outputs values for block, msg and tx variables that are used in the outermost function of a transaction in the counterexample. `msg.value` is also added if the function is payable. These are not reported for internal functions and for external calls to `this`.

### CODE REVIEW

File: libsolidity/formal/CHC.cpp [hunk 1]
```diff
@@ -1890,7 +1890,10 @@ optional<string> CHC::generateCounterexample(CHCSolverInterface::CexGraph const&
 			if (!pred->isConstructorSummary())
 				for (unsigned v: callGraph[node])
 					_dfs(node, v, depth + 1, _dfs);
-			calls.push_front(string(depth * 4, ' ') + pred->formatSummaryCall(nodeArgs(node), m_charStreamProvider));
+
+			bool appendTxVars = pred->isConstructorSummary() || pred->isFunctionSummary() || pred->isExternalCallUntrusted();
```
Comment (leonardoalt) [inline libsolidity/formal/CHC.cpp:1896]: We only want to add the values for external calls to the contract.

File: libsolidity/formal/Predicate.cpp [hunk 2]
```diff
@@ -214,19 +215,64 @@ string Predicate::formatSummaryCall(
...
+	string txModel;
+
+	if (_appendTxVars)
 	{
-		bigint x(*v);
-		if (x > 0)
-			value = "{ value: " + *v + " }";
+		set<string> txVars;
+		if (isFunctionSummary() && programFunction()->isPayable())
+			txVars.insert("msg.value");
```
Comment (cameel) [inline libsolidity/formal/Predicate.cpp]: So the assumption is that `programFunction()` is not NULL when it's a normal function but it could be NULL if it's a constructor? I think there should be an assert against this:
```suggestion
		if (isFunctionSummary())
		{
			smtAssert(programFunction(), "");
			if (programFunction()->isPayable())
				txVars.insert("msg.value");
		}
```

File: libsolidity/formal/Predicate.cpp [hunk 3]
```diff
@@ -214,19 +215,64 @@ string Predicate::formatSummaryCall(
...
+		}
+
+		struct TxVarsVisitor: public ASTConstVisitor
+		{
+			bool visit(MemberAccess const& _memberAccess)
+			{
+				Expression const* memberExpr = SMTEncoder::innermostTuple(_memberAccess.expression());
+
+				auto const& exprType = memberExpr->annotation().type;
+				solAssert(exprType, "");
```
Comment (cameel) [inline libsolidity/formal/Predicate.cpp:245]: I see that we have a special assert type for SMTChecker. This code should be using it then :)
```suggestion
				smtAssert(exprType, "");
```

File: libsolidity/formal/Predicate.cpp [hunk 4]
```diff
@@ -214,19 +215,64 @@ string Predicate::formatSummaryCall(
...
 	{
-		bigint x(*v);
-		if (x > 0)
-			value = "{ value: " + *v + " }";
+		set<string> txVars;
+		if (isFunctionSummary() && programFunction()->isPayable())
+			txVars.insert("msg.value");
+		else if (isConstructorSummary())
+		{
+			auto fun = programFunction();
```
Comment (cameel) [inline libsolidity/formal/Predicate.cpp]: I had to check what kind of thing `programFunction()` is actually returning (wasn't sure if plain `FunctionDefinition` or some wrapper and if it was a pointer or a reference). I think the code is easier to read with more types used explicitly.
```suggestion
			FunctionDefinition const* fun = programFunction();
```

**Comment by leonardoalt** [comment]
This is ready for review.

File: libsolidity/formal/Predicate.cpp [hunk 5]
```diff
@@ -214,19 +215,64 @@ string Predicate::formatSummaryCall(
...
+				txVars.insert("msg.value");
+		}
+
+		struct TxVarsVisitor: public ASTConstVisitor
+		{
+			bool visit(MemberAccess const& _memberAccess)
+			{
+				Expression const* memberExpr = SMTEncoder::innermostTuple(_memberAccess.expression());
+
+				auto const& exprType = memberExpr->annotation().type;
```
Comment (cameel) [inline libsolidity/formal/Predicate.cpp]: I think you're creating a reference to a pointer here :)
```suggestion
				Type const* exprType = memberExpr->annotation().type;
```

File: libsolidity/formal/Predicate.cpp [hunk 6]
```diff
@@ -214,19 +215,64 @@ string Predicate::formatSummaryCall(
...
+				solAssert(exprType, "");
+				if (exprType->category() == Type::Category::Magic)
+				{
+					if (auto const* identifier = dynamic_cast<Identifier const*>(memberExpr))
+					{
+						auto const& name = identifier->name();
+						if (name == "block" || name == "msg" || name == "tx")
+							txVars.insert(name + "." + _memberAccess.memberName());
+					}
+				}
```
Comment (cameel) [inline libsolidity/formal/Predicate.cpp]:
```suggestion
				if (exprType->category() == Type::Category::Magic)
					if (auto const* identifier = dynamic_cast<Identifier const*>(memberExpr))
					{
						auto const& name = identifier->name();
						if (name == "block" || name == "msg" || name == "tx")
							txVars.insert(name + "." + _memberAccess.memberName());
					}
```

File: libsolidity/formal/Predicate.cpp [hunk 7]
```diff
@@ -214,19 +215,64 @@ string Predicate::formatSummaryCall(
...
+			{
+				Expression const* memberExpr = SMTEncoder::innermostTuple(_memberAccess.expression());
+
+				auto const& exprType = memberExpr->annotation().type;
+				solAssert(exprType, "");
+				if (exprType->category() == Type::Category::Magic)
+				{
+					if (auto const* identifier = dynamic_cast<Identifier const*>(memberExpr))
+					{
+						auto const& name = identifier->name();
```
Comment (cameel) [inline libsolidity/formal/Predicate.cpp]:
```suggestion
						ASTString const& name = identifier->name();
```

File: libsolidity/formal/Predicate.cpp [hunk 8]
```diff
@@ -214,19 +215,64 @@ string Predicate::formatSummaryCall(
...
+						if (name == "block" || name == "msg" || name == "tx")
+							txVars.insert(name + "." + _memberAccess.memberName());
```
Comment (cameel) [inline libsolidity/formal/Predicate.cpp:251]: Does this handle stuff like `(msg).value`? Seems to be valid syntax.

Comment (cameel) [inline libsolidity/formal/Predicate.cpp:251]: Shadowing `msg`, `tx` and `block` seems to be only a warning. How will this code deal with something like this?

```solidity
contract C {
    struct S {
        uint value;
        address origin;
        uint number;
    }
    function f() public payable {
        S memory msg = S(42, address(0), 666);
        S memory tx = S(42, address(0), 666);
        S memory block = S(42, address(0), 666);
        msg.value;
        tx.origin;
        block.number;
    }
}
```

File: libsolidity/formal/Predicate.cpp [hunk 9]
```diff
@@ -214,19 +215,64 @@ string Predicate::formatSummaryCall(
...
+			set<string> txVars;
+		} txVarsVisitor;
+
+		if (auto fun = programFunction())
+		{
+			fun->accept(txVarsVisitor);
+			txVars += txVarsVisitor.txVars;
+		}
+
+		auto txValues = readTxVars(_args.at(4));
```
Comment (cameel) [inline libsolidity/formal/Predicate.cpp:267]: Does 4 here mean the argument from function summary at position 4, i.e. `txData`?

File: test/libsolidity/SMTCheckerTest.cpp [hunk 10]
```diff
@@ -66,6 +66,12 @@ SMTCheckerTest::SMTCheckerTest(string const& _filename): SyntaxTest(_filename, E
 		m_ignoreCex = true;
 	else
 		BOOST_THROW_EXCEPTION(runtime_error("Invalid SMT counterexample choice."));
+
+	[[maybe_unused]] auto const& os = m_reader.stringSetting("SMTIgnoreOS", "none");
+#ifdef __APPLE__
+	if (os == "osx")
+		m_shouldRun = false;
+#endif
```
Comment (cameel) [inline test/libsolidity/SMTCheckerTest.cpp]: So this PR also partially implements #11870? :)

Why not add Linux and Windows as an option and extract it into a separate PR?

**Review by cameel — COMMENTED** [review]
I haven't reviewed this fully yet but here's some general feedback.

File: libsolidity/formal/Predicate.cpp (hunk 8 above)
Comment (cameel) [reply libsolidity/formal/Predicate.cpp:251]: Ah, I see it will only run for `Type::Category::Magic`. I'd still add a test that has this kind of shadowing to verify that.

Comment (cameel) [reply libsolidity/formal/Predicate.cpp:251]: Or how about this :)
```solidity
(true ? msg : msg).value;
```

Comment (cameel) [reply libsolidity/formal/Predicate.cpp:251]: Fortunately something like `[msg, tx, block][i].value` is an error though :)

File: test/libsolidity/SMTCheckerTest.cpp (hunk 10 above)
Comment (leonardoalt) [reply test/libsolidity/SMTCheckerTest.cpp]: Yea could do that, I did it here mostly because it was quick and I wanted to see if it would make the tests pass.

File: libsolidity/formal/Predicate.cpp (hunk 9 above)
Comment (leonardoalt) [reply libsolidity/formal/Predicate.cpp:267]: Yep, will add a comment.

File: libsolidity/formal/Predicate.cpp (hunk 8 above)
Comment (leonardoalt) [reply libsolidity/formal/Predicate.cpp:251]: Ok!

Comment (leonardoalt) [reply libsolidity/formal/Predicate.cpp:251]: sigh...

File: libsolidity/formal/Predicate.cpp (hunk 3 above)
Comment (leonardoalt) [reply libsolidity/formal/Predicate.cpp:245]: `smtAssert` is used only in `libsmtutil` I think. We use `solAssert` here because this is all part of `libsolidity`.

File: libsolidity/formal/Predicate.cpp (hunk 2 above)
Comment (leonardoalt) [reply libsolidity/formal/Predicate.cpp]: If it's an actual constructor it should be the same as a normal function, but if it's a contract and that contract has a constructor this should also be true.

Comment (leonardoalt) [reply libsolidity/formal/Predicate.cpp]: Added assertion.

File: libsolidity/formal/Predicate.cpp (hunk 8 above)
Comment (leonardoalt) [reply libsolidity/formal/Predicate.cpp:251]: Added test

Comment (leonardoalt) [reply libsolidity/formal/Predicate.cpp:251]: This case was actually buggy already in develop, nice find :)
Fixed in https://github.com/ethereum/solidity/pull/12036

**Comment by leonardoalt** [comment]
Rebased and ready for review

File: Changelog.md [hunk 11]
```diff
@@ -5,13 +5,13 @@ Language Features:
 
 Compiler Features:
 
+Compiler Features:
```
Comment (chriseth) [inline Changelog.md]: duplication

Comment (leonardoalt) [reply Changelog.md]: fixed

File: Changelog.md [hunk 12]
```diff
@@ -23,7 +21,6 @@ Bugfixes:
  * AST: Export ``canonicalName`` for ``UserDefinedValueTypeDefinition`` and ``ContractDefinition``.
 
 
-
```
Comment (hrkrshnn) [inline Changelog.md]: Some cleaning up needed.

File: test/libsolidity/smtCheckerTests/special/msg_vars_chc_internal.sol [hunk 13]
```diff
@@ -21,8 +21,11 @@ contract C {
...
 		assert(sig == msg.sig); // should hold with CHC
 		assert(value == msg.value); // should hold with CHC
+
+		assert(msg.value == 10); // should fail
 	}
 }
 // ====
 // SMTEngine: chc
 // ----
+// Warning 6328: (621-644): CHC: Assertion violation happens here.\nCounterexample:\ndata = [38, 18, 31, 240], sender = 0, sig = 638722032, value = 0\n\nTransaction trace:\nC.constructor()\nState: data = [], sender = 0, sig = 0, value = 0\nC.f(){ msg.data: [38, 18, 31, 240], msg.sender: 0, msg.sig: 638722032, msg.value: 0 }\n    C.g() -- internal call
```
Comment (hrkrshnn) [inline test/libsolidity/smtCheckerTests/special/msg_vars_chc_internal.sol]: For data or sig, I think hex is a better choice. But fine either way.

File: libsolidity/formal/Predicate.cpp [hunk 14]
```diff
@@ -214,19 +215,67 @@ string Predicate::formatSummaryCall(
...
+
+		// Here we are interested in txData from the summary predicate.
+		auto txValues = readTxVars(_args.at(4));
+		vector<string> values;
+		for (auto const& _var: txVars)
+			if (auto v = txValues.at(_var))
+				values.push_back(_var + ": " + *v);
+
+		if (!values.empty())
+			txModel = "{ " + boost::algorithm::join(values, ", ") + " }";
```
Comment (hrkrshnn) [inline libsolidity/formal/Predicate.cpp:274]: I wonder if replacing this by the Json library is a good idea?

File: Changelog.md (hunk 12 above)
Comment (leonardoalt) [reply Changelog.md]: This is weird, I don't have it in my local diff

File: test/libsolidity/smtCheckerTests/special/msg_vars_chc_internal.sol (hunk 13 above)
Comment (leonardoalt) [reply test/libsolidity/smtCheckerTests/special/msg_vars_chc_internal.sol]: I think so too. Let me see if I can do this in an easy way, otherwise I'd vote to do it in a different PR.

File: libsolidity/formal/Predicate.cpp (hunk 14 above)
Comment (leonardoalt) [reply libsolidity/formal/Predicate.cpp:274]: which json library?

Comment (hrkrshnn) [reply libsolidity/formal/Predicate.cpp:274]: Just the regular `json-cpp` that we use. `txModel` would be `Json::value` and then you can write `txModel["msg.data"] = data`.

Comment (leonardoalt) [reply libsolidity/formal/Predicate.cpp:274]: But how does that help?

Comment (hrkrshnn) [reply libsolidity/formal/Predicate.cpp:274]: No need to a write custom translator to strings. The final output would be just the JSON as string.

Comment (leonardoalt) [reply libsolidity/formal/Predicate.cpp:274]: Oh I see what you mean, for the entire message?

Comment (hrkrshnn) [reply libsolidity/formal/Predicate.cpp:274]: Yes!

File: libsolidity/formal/Predicate.cpp [hunk 15]
```diff
@@ -214,19 +219,79 @@ string Predicate::formatSummaryCall(
...
+		// We can either have a pretty version which has line breaks,
+		// or a compact version that does not but also has no spaces
+		// and has quotes.
+		// So we take the pretty version and remove line breaks,
+		// replace double spaces (which are left after removing line breaks)
+		// by single spaces, and remove quotes.
+		txModelStr = util::jsonPrettyPrint(txModel);
+		boost::replace_all(txModelStr, "\n", "");
+		boost::replace_all(txModelStr, "  ", " ");
+		boost::replace_all(txModelStr, "\"", "");
```
Comment (leonardoalt) [inline libsolidity/formal/Predicate.cpp]: @hrkrshnn this gives me what I want. Let me know if you prefer this or the previous version

Comment (leonardoalt) [reply libsolidity/formal/Predicate.cpp]: Actually it doesn't. The end of the string is then `name: value}`, whereas it should be `name: value }`. It's getting too hacky, I'm reverting

File: libsolidity/formal/Predicate.cpp [hunk 16]
```diff
@@ -575,9 +595,9 @@ map<string, optional<string>> Predicate::readTxVars(smtutil::Expression const& _
 		{"block.number", TypeProvider::uint256()},
 		{"block.timestamp", TypeProvider::uint256()},
 		{"blockhash", TypeProvider::array(DataLocation::Memory, TypeProvider::uint256())},
-		{"msg.data", TypeProvider::bytesMemory()},
+		{"msg.data", TypeProvider::array(DataLocation::Memory, TypeProvider::fixedBytes(1))},
```
Comment (hrkrshnn) [inline libsolidity/formal/Predicate.cpp]: It is `calldata` in 
https://github.com/ethereum/solidity/blob/9d6eaa7a46425301ddb961a369b04f211d53691c/libsolidity/ast/Types.cpp#L3963

But not sure if that would be a problem here.

Maybe the `fixedBytes` argument can also be avoided?

File: Changelog.md [hunk 17]
```diff
@@ -10,6 +10,8 @@ Bugfixes:
   * SMTChecker: Fix internal error in magic type access (``block``, ``msg``, ``tx``).
 
 
+
+
```
Comment (hrkrshnn) [inline Changelog.md]: I think this commit can be squashed to the first one.

**Review by hrkrshnn — COMMENTED** [review]
Looks good. Some small comments.

File: Changelog.md (hunk 17 above)
Comment (leonardoalt) [reply Changelog.md]: Yea I'm gonna squash everything after review

File: libsolidity/formal/Predicate.cpp (hunk 16 above)
Comment (leonardoalt) [reply libsolidity/formal/Predicate.cpp]: Would that make the element type a `fixedBytes(1)` implicitly?

Comment (leonardoalt) [reply libsolidity/formal/Predicate.cpp]: We need that to report the type properly as hex.

Comment (hrkrshnn) [reply libsolidity/formal/Predicate.cpp]: Yes, `fixedBytes(1)` is implicit.

Comment (leonardoalt) [reply libsolidity/formal/Predicate.cpp]: Ok, changed and squashed.

**Review by hrkrshnn — APPROVED** [review]
